### PR TITLE
[LIB:DEBUGKD0] Kernel phase 0 debug library.

### PR DESCRIPTION
**Pull request by vgalnt**
The DbgPrint0() function outputs a debug message to one of the COM ports, which must be previously initialized using DbgKdPortInitialize().

Purpose:
In the Virtual Box virtual machine, it is possible to output to the COM port in phase 0 of the ReactOS kernel initialization (even when interrupts are disabled). This way we can debug the kernel, HAL and drivers in phase 0.
This is disabled by default with DBG_KD0 for both build types ("Release" and "Debug").

### CODE REVIEW

File: sdk/lib/drivers/debugkd0/debugkd0.c [hunk 1]
```diff
@@ -0,0 +1,173 @@
+/*
+ * PROJECT:         ReactOS Kernel
+ * COPYRIGHT:       GPL-2.0+ (https://spdx.org/licenses/GPL-2.0+)
+ * FILE:            lib/drivers/debugkd0/debugkd0.c
+ * PURPOSE:         Kernel phase 0 debug library header
```
Comment (SergeGautherie) [inline sdk/lib/drivers/debugkd0/debugkd0.c:5]: `header`: copypasta...

File: sdk/lib/drivers/debugkd0/debugkd0.h [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+BOOLEAN
+NTAPI
+DbgKdPortInitialize(_In_ ULONG PortNumber OPTIONAL,
+                    _In_ PUCHAR PortAddress OPTIONAL,
+                    _In_ ULONG BaudRate OPTIONAL,
+                    _Out_ PULONG OutPortId);
+
+#if DBG
+
+  #if DBG_KD0
```
[the remark this answers is not in the document]
Comment (vgalnt) [reply sdk/lib/drivers/debugkd0/debugkd0.h:41]: Phase 0 debugging is rarely needed, so DBG_KD0 defaults to 0.
DBG and DBG_KD0 are used for different purposes.
You can have DEBUGKD0 disabled while DBG is enabled and you can have DEBUGKD0 enabled while DBG is disabled.

File: sdk/lib/drivers/debugkd0/debugkd0.c (hunk 1 above)
Comment (vgalnt) [reply sdk/lib/drivers/debugkd0/debugkd0.c:5]: Yes. Thank you.

File: sdk/lib/drivers/debugkd0/debugkd0.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (SergeGautherie) [reply sdk/lib/drivers/debugkd0/debugkd0.h:41]: > you can have DEBUGKD0 enabled while DBG is disabled

That confirms `DBG` must not be checked here, doesn't it?

[the remark this answers is not in the document]
Comment (vgalnt) [reply sdk/lib/drivers/debugkd0/debugkd0.h:41]: These are the details that each developer (who needs debugging in phase 0) can modify to fit their needs.

[the remark this answers is not in the document]
Comment (SergeGautherie) [reply sdk/lib/drivers/debugkd0/debugkd0.h:41]: I don't understand your answer, sorry:
if that very code is what you do want here, then I do not understand the block at line 13 and I do not understand why .c does not have the same `#if`s as here...

[the remark this answers is not in the document]
Comment (vgalnt) [reply sdk/lib/drivers/debugkd0/debugkd0.h:41]: If you use the debugkd0 library and block at line 13 bothers you, you can use, for example, only one line instead:
#define DBG_KD0 1

This will change depending on the needs of the developer. These are details that are not important in this case. This code will be used very rarely. And only for debugging.

**Comment by Extravert-ir** [comment]
For early debugging, I use a Timo's trick with passing Freeloader's `DbgPrint` to the kernel via `LoaderBlock`.
Does this method have any benefit over it?
https://github.com/reactos/reactos/blob/bb209e9010ecddfe89150261b937ea8c73c1c823/boot/freeldr/freeldr/ntldr/arch/amd64/winldr.c#L385
https://github.com/reactos/reactos/blob/bb209e9010ecddfe89150261b937ea8c73c1c823/ntoskrnl/ke/amd64/kiinit.c#L380

File: sdk/lib/drivers/debugkd0/debugkd0.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (SergeGautherie) [reply sdk/lib/drivers/debugkd0/debugkd0.h:41]: > only one line instead

Yes, that would be one way.

>  not important

I don't understand why you like these inconsistencies and I don't understand how how often this is used matters.
Anyway, proceed as you like...

**Comment by vgalnt** [comment]
DbgPrint and DbgPrint0 use different mechanisms (INT trap and simple COM port output respectively).
DbgPrint() is available only after initializing the debugger. Before that you will not be able to use it. As well as at the time of reinitialization processor (trap vectors) and debbugger. My library can always be used (at least in the Virtual Box) from any location of the codes.

**Comment by Extravert-ir** [comment]
I was talking about __FreeLoader's DbgPrint__, it's working same way as your `DbgPrint0` (with additional Bochs support)
https://github.com/reactos/reactos/blob/master/boot/freeldr/freeldr/lib/debug.c#L199-L219

**Comment by vgalnt** [comment]
Yes, I mistakenly thought that the ReactOS FreeLoader uses the same mechanism as NT.
But I felt at least a little like the creator of the bicycle :)

**Comment by vgalnt** [comment]
Closed

**Comment by vgalnt** [comment]
Closed

**Comment by vgalnt** [comment]
My browser does not support, please close.
